[NativePromise] Shouldn't require to copy the resolved value when the promise isn't exclusive

jyavenard opened the pull request:
#### c1b873c21745249657200c06d205ee9ae31c7692
<pre>
[NativePromise] Shouldn&apos;t require to copy the resolved value when the promise isn&apos;t exclusive
<a href="https://bugs.webkit.org/show_bug.cgi?id=262872">https://bugs.webkit.org/show_bug.cgi?id=262872</a>
rdar://116664254

Reviewed by Youenn Fablet.

In Non-Exclusive mode, use a ref-counted storage that we pass around
rather than copying the data across all completion promises.

Fly-by fix: logging incorrectly returned that the promise had always been
resolved.

* Source/WTF/wtf/NativePromise.h:
* Tools/TestWebKitAPI/Tests/WTF/NativePromise.cpp:
(TestWebKitAPI::TEST):

Canonical link: <a href="https://commits.webkit.org/269241@main">https://commits.webkit.org/269241@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/60ee508de5cbad52924c9a4aa7925102732b3a58

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/21985 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/22201 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/23052 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/23867 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/20356 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/22229 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/26448 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/22513 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/21437 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/22214 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/21861 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/19071 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/24719 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/18978 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/19925 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/26185 "Found 17 new test failures: imported/w3c/web-platform-tests/css/css-masking/clip-path/clip-path-descendant-text-mutated-001.html, imported/w3c/web-platform-tests/css/css-pseudo/backdrop-animate-002.html, imported/w3c/web-platform-tests/css/css-transforms/animation/transform-interpolation-rotate.html, imported/w3c/web-platform-tests/css/css-transforms/group/svg-transform-nested-009.html, imported/w3c/web-platform-tests/css/css-transforms/individual-transform/animation/individual-transform-combine.html, imported/w3c/web-platform-tests/css/css-transforms/transform-3d-rotateY-stair-above-001.xht, imported/w3c/web-platform-tests/css/css-ui/compute-kind-widget-generated/kind-of-widget-fallback-color-input-background-color-001.html, imported/w3c/web-platform-tests/css/css-ui/compute-kind-widget-generated/kind-of-widget-fallback-select-dropdown-box-background-color-001.html, imported/w3c/web-platform-tests/css/css-ui/compute-kind-widget-generated/kind-of-widget-fallback-select-dropdown-box-border-bottom-left-radius-001.html, imported/w3c/web-platform-tests/css/css-ui/compute-kind-widget-generated/kind-of-widget-fallback-select-dropdown-box-border-bottom-right-radius-001.html ... (failure)") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/19113 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/20002 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/20149 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/24054 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/21358 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/20579 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/17529 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/25402 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/19932 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/5993 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/5242 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/24137 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/26675 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/20530 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/5825 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

squelart commented on Source/WTF/wtf/NativePromise.h:
```diff
@@ -368,6 +368,16 @@ class NativePromise final : public NativePromiseBase, public ConvertibleToNative
 
 private:
     friend class Producer;
+    class Storage;
```
Is this forward declaration necessary? I don't see any mention of `Storage` before it's defined at line 1011 below.

jyavenard replied:
yes, I got an error about it.

squelart commented on Source/WTF/wtf/NativePromise.h:
```diff
@@ -368,6 +368,16 @@ class NativePromise final : public NativePromiseBase, public ConvertibleToNative
+    inline void settleImpl(SettleValueType&& result, Locker<Lock>& lock)
+    {
+        assertIsHeld(m_lock);
+        m_result.optionalResult() = std::forward<SettleValueType>(result);
```
Suggestion:
`std::optional::operator=` is quite "wide", is that it can handle lots of cases, like copy-assignment (from the same type of optional); and depending on whether the optional is already engaged it may do either an initialization or an assignment; etc.
But I believe this code here should only be used once, to set a value in an empty optional, right?
So I would suggest using `emplace`, which just focuses on initializing the contained type with the given argument.

youennf replied:
You could move the isNothing() ASSERT here, or make m_result have an initialise method that would ASSERT it is not already initialised.

squelart reviewed (approved):
r+ with optional nits.
(Unofficial WK reviewer, please get a 2nd review.)

youennf commented on Source/WTF/wtf/NativePromise.h:
```diff
@@ -513,11 +519,9 @@ class NativePromise final : public NativePromiseBase, public ConvertibleToNative
+                });
                 m_producer->resolve(WTFMove(results));
                 m_producer = nullptr;
                 m_results.clear();
```
I would be tempted to move m_results before the m_producer->resolve call for extra safety.
It will make sure m_results never gets used in the state where it contains moved items.

jyavenard committed:
[NativePromise] Shouldn't require to copy the resolved value when the promise isn't exclusive
https://bugs.webkit.org/show_bug.cgi?id=262872
rdar://116664254

Reviewed by Youenn Fablet.

In Non-Exclusive mode, use a ref-counted storage that we pass around
rather than copying the data across all completion promises.

Fly-by fix: logging incorrectly returned that the promise had always been
resolved.

* Source/WTF/wtf/NativePromise.h:
* Tools/TestWebKitAPI/Tests/WTF/NativePromise.cpp:
(TestWebKitAPI::TEST):

Canonical link: https://commits.webkit.org/269241@main
